### ISSUE TITLE
[tx] Implement heuristic to determine micro batch sizes automatically

### ISSUE THREAD

**Issue by pcmoritz** [NovaSky-AI/SkyRL#1048]
Currently most users have to set the micro batch sizes for sampling and training manually (`train_micro_batch_size` and `sample_max_num_sequences` respectively) to avoid OOM. We should come up with a good heuristic that sets these values automatically if they are set to `0`.

I'm thinking of something along these lines:

- After allocating the base model and optimizer states, compute the remaining memory on the GPUs
- Based on the metadata of the model (which we can do in tx since we control the model definition), compute how much memory each additional token takes up in terms of activations, separately for sampling and training (the latter will also include the gradients)
- For each micro batch iteration, use the highest batch size that still fits in the available memory, and we could also do some sorting according to sequence length for selecting which sequences to batch, so we maximize utilization.

Ideally we come up with something that is very simple and works well by default, so users don't need to care about setting these parameters by hand.

**Comment by jonahsamost** [comment]
I can take this. On top of the more analytical path to determine micro batch size, what do you think of also providing an optional profiling mode (like run a forward and backward pass, and measure actual memory pressure) and have that determine the micro batch size? Also, should I be building on top of PR #847 or be working against main?

**Comment by pcmoritz** [comment]
So for the Jax backend, which I'm the most familiar with, I think we can get the information from XLA https://github.com/AI-Hypercomputer/maxtext/blob/28e5097ac467ed8b1d17676d68aa5acc50f9d60d/src/MaxText/train.py#L380-L382

That would probably be better than analytically trying to determine the memory (since it depends on the model implementation which is going to evolve, might depend on flags, and also will depend on sharding, and it is probably unreasonable / imprecise to try to model all of these analytically).

For other backends, we should see if pytorch / fsdp or megatron have a way to measure or estimate the memory usage :)

I would work against main, since estimating the memory and using it to set the parameters can be independent and this way there will be less conflicts.

**Comment by jonahsamost** [comment]
Sounds good, will let you know if I have any questions
